cli/src/Api.ts

```diff
@@ -0,0 +1,135 @@
+import BN from 'bn.js';
+import { registerJoystreamTypes } from '@joystream/types';
+import { ApiPromise, WsProvider } from '@polkadot/api';
+import { formatBalance } from '@polkadot/util';
+import { Balance, Hash } from '@polkadot/types/interfaces';
+import { KeyringPair } from '@polkadot/keyring/types';
+import { AccountBalances, CouncilInfoObj, CouncilInfoTuple, AccountSummary, createCouncilInfoObj } from './Types';
+import { DerivedFees, DerivedBalances } from '@polkadot/api-derive/types';
+import { CLIError } from '@oclif/errors';
+import ExitCodes from './ExitCodes';
+
+const API_URL = process.env.WS_URL || (
+    process.env.MAIN_TESTNET ?
+        'wss://rome-rpc-endpoint.joystream.org:9944/'
+        : 'wss://rome-staging-2.joystream.org/staging/rpc/'
+);
+const TOKEN_SYMBOL = 'JOY';
+
+// Api wrapper for handling most common api calls and allowing easy API implementation switch in the future
+
+export default class Api {
+    private _api: ApiPromise | null = null;
+    private initializing: boolean = false;
+
+    private async initApi(): Promise<ApiPromise> {
+        formatBalance.setDefaults({ unit: TOKEN_SYMBOL });
+        const wsProvider:WsProvider = new WsProvider(API_URL);
+        registerJoystreamTypes();
+
+        return await ApiPromise.create({ provider: wsProvider });
+    }
+
+    private async getApi(): Promise<ApiPromise> {
+        while (this.initializing) {
```

Comment:
What is going on here? This construction looks very suspicious? If this was lifted from Pioneer, then its probably just because Pioneer is badly constructed. Let me know if I am missing something.

Reply:
I refactored this class a few times, so that's why it may now seem kind of weird.
The idea was that it should prevent accidentialy initializing the api twice. So if the api initialization was already started somewhere, it should then await 10 ms intervals until the the initialization process has finished. So in case we do something like:

```
const api: Api = new Api();
api.getCouncilInfo(); // Async call, initializes the api if not yet initialized
api.getAccountSummary(SOME_ACCOUNT); // Same as above
```

The second call should not re-initialize the api, but instead wait until the first call finishes the initialization.

This made a little more sense when this class was still a singleton, later I decided to change it into something more like the original ApiPromise, so:

```
const api = await Api.create();
// Now we can just use the api knowing that it has already been initialized.
// It may also throw an error if we try to use it before initialization
```

But I probably kind of confused myself during the process and ended up with some hybrid of two potential approches. I'll fix it, by going for the second one, as it seems more "clean".